reconcilers/childset.go

```diff
@@ -0,0 +1,359 @@
+	//
+	// Defaults to filtering by the reconciled resource's namespace:
+	//     []client.ListOption{
+	//         client.InNamespace(resource.GetNamespace()),
+	//     }
+	//
+	// +optional
+	ListOptions func(ctx context.Context, resource Type) []client.ListOption
+
+	// OurChild is used when there are multiple ChildSetReconciler for the same ChildType controlled
+	// by the same reconciled resource. The function return true for child resources managed by
+	// this ChildSetReconciler. Objects returned from the DesiredChildren function should match this
+	// function, otherwise they may be orphaned.
+	//
+	// A stable identifier for the child resource must be returned when the resource is a child
+	// controlled by this reconciler. The identifier is used to correlate desired child resources
+	// with actual child resources. Non-deterministic IDs will result in the rapid deletion and
+	// creation of child resources.
+	OurChild func(resource Type, child ChildType) (bool, string)
+
+	// Sanitize is called with an object before logging the value. Any value may
+	// be returned. A meaningful subset of the resource is typically returned,
+	// like the Spec.
+	//
+	// +optional
+	Sanitize func(child ChildType) interface{}
+
+	stamp          *ResourceManager[ChildType]
+	lazyInit       sync.Once
+	voidReconciler *ChildReconciler[Type, ChildType, ChildListType]
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) init() {
+	r.lazyInit.Do(func() {
+		var nilCT CT
+		if internal.IsNil(r.ChildType) {
+			r.ChildType = newEmpty(nilCT).(CT)
+		}
+		if internal.IsNil(r.ChildListType) {
+			var nilCLT CLT
+			r.ChildListType = newEmpty(nilCLT).(CLT)
+		}
+		if r.Name == "" {
+			r.Name = fmt.Sprintf("%sChildSetReconciler", typeName(r.ChildType))
+		}
+		r.stamp = &ResourceManager[CT]{
+			Name:                     r.Name,
+			Type:                     r.ChildType,
+			Finalizer:                r.Finalizer,
+			TrackDesired:             r.SkipOwnerReference,
+			HarmonizeImmutableFields: r.HarmonizeImmutableFields,
+			MergeBeforeUpdate:        r.MergeBeforeUpdate,
+			Sanitize:                 r.Sanitize,
+		}
+		r.voidReconciler = r.childReconcilerFor(nilCT, nil, "")
+	})
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) SetupWithManager(ctx context.Context, mgr ctrl.Manager, bldr *builder.Builder) error {
+	r.init()
+
+	c := RetrieveConfigOrDie(ctx)
+
+	log := logr.FromContextOrDiscard(ctx).
+		WithName(r.Name).
+		WithValues("childType", gvk(r.ChildType, c.Scheme()))
+	ctx = logr.NewContext(ctx, log)
+
+	if err := r.validate(ctx); err != nil {
+		return err
+	}
+
+	if err := r.voidReconciler.SetupWithManager(ctx, mgr, bldr); err != nil {
+		return err
+	}
+
+	if r.Setup == nil {
+		return nil
+	}
+	return r.Setup(ctx, mgr, bldr)
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) childReconcilerFor(desired CT, desiredErr error, id string) *ChildReconciler[T, CT, CLT] {
+	return &ChildReconciler[T, CT, CLT]{
+		Name:               id,
```

Comment:
Would it be better to name this with `r.Name + "-" + id`? In our use case we have two ChildSetReconcilers generating different sets of child resources based on the same set of ids. Just using id can be problematic and harder to read in the logs.

Reply:
the logr name is hierarchical, so the effective name the appears in log lines already contains the name of the ChildSetReconciler.